polyply/src/annotate_dna.py

```diff
@@ -0,0 +1,194 @@
+        """
+
+        def _is_nucleobase(node):
+            return (meta_molecule.nodes[node]['resname'] in base_library)
+
+        seperated_meta_mol = [meta_molecule.subgraph(graph) for graph
+                              in nx.connected_components(meta_molecule)]
+
+        # Find all strands by performing greedy search on
+        # all connected_components of meta_molecule
+        strands = []
+        for meta_mol in seperated_meta_mol:
+            queue = sample(meta_mol.nodes, 1)
+            visited = []
+            strand = []
+            while queue:
+                # pop first node out of queue
+                current_node = queue.pop()
+                visited.append(current_node)
+
+                if _is_nucleobase(current_node):
+                    strand.append(current_node)
+                elif strand:
+                    strands.append(meta_mol.subgraph(strand))
+                    strand = []
+
+                # Determine neighbors nodes
+                neighbors = [node for node in meta_molecule.neighbors(current_node) if node not in queue + visited]
+                # update queue
+                queue = neighbors + queue
+                # Sort queue according to heuristic
+                queue.sort(key=_is_nucleobase, reverse=True)
+
+            if strand not in strands:
+                strands.append(meta_molecule.subgraph(strand))
+        return strands
+
+    def _determine_sequence(self, strand):
+        sequence = []
+        if nx.cycle_basis(strand):
+            start_base = sample(strand.nodes, 1)
+        else:
+            condition = lambda node: strand.degree(node) == 1
+            start_base = next(filter(condition, strand.nodes), None)
+        for resid in nx.dfs_tree(strand, start_base):
+            sequence.append(strand.nodes[resid]['resname'])
+        return sequence
+
+    def _find_complementary_strands(self, ref_strand, strands):
+        # Sequence we want to find match for
+        ref_seq = self._determine_sequence(ref_strand)
+
+        match_ratios = {}
+        for ndx, strand in enumerate(strands, start=1):
+            seq = self._determine_sequence(strand)
+
+            match_ratio = 0
+            for ref_base, base in zip(ref_seq, seq):
+                if base_library[ref_base] == base:
+                    match_ratio += 1
+            match_ratios[ndx] = match_ratio/len(ref_seq)
+
+            match_ratio = 0
+            for ref_base, base in zip(ref_seq, reversed(seq)):
+                if base_library[ref_base] == base:
+                    match_ratio += 1
+            match_ratios[-ndx] = match_ratio/len(ref_seq)
+
+        match =  max(match_ratios, key=match_ratios.get)
+        match_strand = strands.pop(abs(match) - 1)
+
+        if match <= 0:
+            mapping = dict(zip(match_strand, sorted(match_strand, reverse=True)))
+            match_strand = nx.relabel_nodes(match_strand, mapping)
```

Comment:
Why this?

Reply:
When 'match < 0' the optimal sequence match is found by reversing one of the strands. This mapping reverses the labels of that specific strand so that when '_combine_complementary_strands' loops through the sorted strand, it effectively loops through the strand in reverse. I do it this way, since I found no obvious way to iterate over a undirected network in reverse.  There probably is however a more efficient way, I am thinking about it.